archunit-integration-test/src/test/java/com/tngtech/archunit/ImporterRules.java

```diff
@@ -26,6 +26,7 @@
                     .that().resideOutsideOfPackage(THIRDPARTY_PACKAGE_IDENTIFIER)
                     .and(not(belongToAnyOf(JavaType.class)))
                     .and().doNotHaveFullyQualifiedName("com.tngtech.archunit.core.importer.JavaTypeImporter")
+                    .and().doNotHaveFullyQualifiedName("com.tngtech.archunit.core.importer.JavaClassProcessor$AnnotationArrayProcessor")
```

Comment:
This should not be necessary, and actually it is an intended rule :wink: If it would really fail it would mean you're actually not doing it the way intended. In the past I did have some trouble when I found corner cases where `..asm..Type` would not behave the way I needed it to. So I encapsulated all that knowledge within `JavaTypeImporter` and forbid direct access to `..asm..Type` so I would not run into the same pitfall again :wink:

Reply:
I thought that it is there for a reason and I was confused myself why this test failed. But `JavaClassProcessor` was always using `..asm..Type`. It is just with this change that it becomes apparent, because `instanceOf Type` is checked :D

Reply:
Oh yes, I remember seeing this and thinking "not so much harm and not detected anyway at the moment" :joy: Because what I was worried about couldn't happen by an `instanceof` check :wink: I think even this thing should then probably be encapsulated in `JavaTypeImporter` or `JavaType`. I really want to avoid as many dependencies on `..asm..Type` outside of these classes as possible as it is a source of error as mentioned...